### ISSUE TITLE
Races detected in visitor code

### ISSUE THREAD

**Issue by owenniles** [lovoo/goka#461]
I am running a Kafka cluster with three brokers. When I run the visitor integration tests with race detection enabled, the tests fail because races are detected in two of the test cases.
```
~/goka $ GOKA_SYSTEMTEST= GOKA_SYSTEMTEST_BROKERS=localhost:9091,localhost:9092,localhost:9093 go test ./systemtest -run TestProcessorVisit -race
(verbose errors omitted)
--- FAIL: TestProcessorVisit (37.64s)
    --- FAIL: TestProcessorVisit/visit-panic-slow (4.36s)
        testing.go:1398: race detected during execution of test
    --- FAIL: TestProcessorVisit/visit-shutdown (5.96s)
        testing.go:1398: race detected during execution of test
FAIL
FAIL    github.com/lovoo/goka/systemtest        37.898s
FAIL
```
I have confirmed that this behavior exists both on the latest version of `master` (183abf8316af159c83f8d1e538e5599fafdf64ea) and v1.1.12.

I'm raising this issue because I believe the races may be affecting my production code. I'm observing panics with stack traces identical to the one in #433. I noticed that the races detector is detecting races in the same place. I will create a separate ticket for my specific case if I'm able to isolate the issue further.

**Comment by abligh** [comment]
Independent reproduction of this, plus a root cause and a fix — reproducible **without a Kafka cluster**, in under a second.

We hit this downstream as a rare `panic: send on closed channel` during shutdown, with a stack identical to #433's. Chasing it led back here, so I am adding what we found rather than opening a duplicate.

### Reproducer (no brokers, uses `goka/tester`)

Drop into `integrationtest/`, `go test -race -run TestVisitRacesShutdown ./integrationtest`:

```go
func TestVisitRacesShutdown(t *testing.T) {
	const (
		iterations = 200
		keys       = 200
	)

	for i := 0; i < iterations; i++ {
		gkt := tester.New(t)

		proc, err := goka.NewProcessor(nil,
			goka.DefineGroup("visit-race",
				goka.Input(goka.Stream("input"), new(codec.String), func(ctx goka.Context, msg any) {
					ctx.SetValue(msg)
				}),
				goka.Visitor("visitor", func(ctx goka.Context, meta any) {
					ctx.SetValue(ctx.Value())
				}),
				goka.Persist(new(codec.String)),
			),
			goka.WithTester(gkt),
		)
		require.NoError(t, err)

		ctx, cancel := context.WithCancel(context.Background())
		procDone := make(chan struct{})
		go func() {
			defer close(procDone)
			_ = proc.Run(ctx)
		}()
		require.NoError(t, proc.WaitForReadyContext(ctx))

		for k := 0; k < keys; k++ {
			gkt.Consume("input", fmt.Sprintf("key-%d", k), "value")
		}

		visitDone := make(chan struct{})
		go func() {
			defer close(visitDone)
			_, _ = proc.VisitAllWithStats(ctx, "visitor", nil)
		}()
		cancel()

		<-visitDone
		<-procDone
	}
}
```

It panics within the first handful of iterations, typically in about 0.3s, **with or without `-race`**:

```
panic: send on closed channel

goroutine 1223 [running]:
github.com/lovoo/goka.(*PartitionProcessor).VisitValues(...)
	github.com/lovoo/goka@v1.1.17/partition_processor.go:673 +0x418
github.com/lovoo/goka.(*Processor).VisitAllWithStats.func1()
	github.com/lovoo/goka@v1.1.17/processor.go:977 +0x2c
github.com/lovoo/goka.(*Processor).VisitAllWithStats.(*ErrGroup).Go.func2()
	github.com/lovoo/goka@v1.1.17/multierr/errgroup.go:48 +0x38
```

Under `-race` the detector sometimes gets a report out before the panic kills the process, and when it does it is one of two, both between the same pair of goroutines (a `VisitAllWithStats` fan-out goroutine and a `Cleanup` → `Stop` one):

* more often, `multierr.(*ErrGroup).Go` at `multierr/errgroup.go:47`, called from `VisitValues` at `partition_processor.go:659`, racing `(*ErrGroup).Wait` at `multierr/errgroup.go:28`, called from `Stop` at `partition_processor.go:305`;
* less often, the send at `partition_processor.go:683` racing `Stop` at `partition_processor.go:308` — i.e. the send and `close(pp.visitInput)`.

The first of those is the interesting one, and it is the one this fix is aimed at — see below.

This is not a `tester` artifact (#481 is a genuinely separate race, inside `tester/consumergroup.go`): every frame here is in goka core, and `tester` only supplies the ordinary `Setup`/`Cleanup` lifecycle that a real broker also drives on every rebalance.

### Root cause — there are two layers, and the outer one is a decoy

**Layer 1, the visible panic.** `VisitValues` enqueues each key with a `select` over `<-stopping`, `<-ctx.Done()` and `pp.visitInput <- &visit{...}`. `Stop()` does set `PPStateStopping` (closing `stopping`) well before it reaches `close(pp.visitInput)` — but `select` gives no priority to a ready receive over a ready send, and a send on a closed channel is always "ready": it panics the moment it is chosen. So once `Stop()` has closed the channel, a concurrent `VisitValues` sitting in that `select` panics instead of returning `ErrVisitAborted`.

**Layer 2, the actual defect.** The only thing meant to prevent that overlap is `VisitValues` registering `pp.runnerGroup.Go(func() error { <-visitDone; return nil })`, with `Stop()` blocking on `pp.runnerGroup.Wait()` before it closes anything. But nothing orders that registration against `Stop()`. `VisitAllWithStats` spawns the per-partition `VisitValues` goroutines under `g.mTables.RLock()` and releases the lock immediately — before any of them has run a line — so a `Cleanup()` (every rebalance, not just shutdown) can already be inside `Stop()` for that partition. Guarding only the channel close makes this visible: I tried that first, and the failure simply moved to

```
panic: sync: WaitGroup is reused before previous Wait has returned
```

with `(*ErrGroup).Go` from `VisitValues` racing `(*ErrGroup).Wait` from `Stop` — which I think is the same thing #433 was seeing as "negative WaitGroup counter". The fix has to make joining `runnerGroup` and waiting on it mutually exclusive; the channel-close race then closes on its own, because `Stop()` only closes after a `Wait()` that now genuinely covers every visitor.

### Fix

PR: #493. It adds a mutex on `PartitionProcessor` that `VisitValues` takes to join `runnerGroup` and `Stop()` takes to declare the group closed before waiting; a visitor arriving after that returns `ErrVisitAborted` instead of joining a group already being waited on. The reproducer above ships with it as `integrationtest/visit_shutdown_test.go`, so it runs in normal CI — no broker needed, about 2.7s.

The PR also carries two much smaller, independent bugs found in the same function:

- `VisitAllWithStats` takes `g.mTables.RLock()` and returns on the stateless path **without unlocking it** — any later `mTables.Lock()` (e.g. `Cleanup` resetting `g.partitions`) then blocks forever.
- `VisitAllWithStats` finishes with a plain read of `visited`, which in-flight visits are still incrementing via `atomic.AddInt64` whenever `VisitValues` bailed out early on cancellation without waiting for them. That one is straightforwardly `-race`-detectable on its own.

### What this was tested against

All on go1.26.5, darwin/arm64.

- **v1.1.17** — the current release, and also `master`: the `v1.1.17` tag sits on 60275b5, which is `master` HEAD, so those are one and the same commit rather than two separate results. Dropped into `integrationtest/` on an otherwise unmodified tree, this panics in **5/5 runs** (3 with `-race`, 2 without), each within ~0.4s.
- **v1.1.12** — the version this issue was originally reported against. Same reproducer, same panic, **3/3 runs** under `-race`. So this is not a regression introduced after your report.
- **v1.1.17 + the PR** — `go test -race ./integrationtest` passes 3/3 runs, and `go test -race ./...` passes for every goka package except `systemtest` and `storage/redis`, which I skipped because they need a broker and a Redis respectively.

I have **not** run `systemtest` against real brokers — I have no cluster to hand. If someone can run `TestProcessorVisit` (the `visit-shutdown` / `visit-panic-slow` subtests from this issue) with the patch applied, that is the confirmation I cannot supply.

Happy to split any of it out, rework the approach, or move the reproducer into `systemtest` instead if you would rather see it there.

### AI Disclosure

For full disclosure this Comment, the bug isolation within it, and the proposed fix has been substantially aided by AI, specifically Claude Sonnet 5 and Claude Opus 5. abligh is however a real live human (see other repos), and has reviewed the output. And this bug has been causing issues in real code. If AI use is a problem in this repo, please let me know.
